src/server/pkg/storage/chunk/reader.go

```diff
@@ -58,20 +66,27 @@ func (r *Reader) Read(data []byte) (int, error) {
 		data = data[n:]
 		totalRead += n
 		if err != nil {
-			// If all DataRefs have been read, then io.EOF.
-			if len(r.dataRefs) == 0 {
-				return totalRead, io.EOF
-			}
 			if err := r.nextDataRef(); err != nil {
 				return totalRead, err
 			}
 		}
 	}
 	return totalRead, nil
-
 }
 
 func (r *Reader) nextDataRef() error {
+	// If all DataRefs have been read, then io.EOF.
+	if len(r.dataRefs) == 0 {
+		if r.f != nil {
+			var err error
+			r.dataRefs, err = r.f()
+			if err != nil {
+				return err
+			}
+		} else {
+			return io.EOF
+		}
+	}
```

Comment:
Is there somewhere that you set `r.f` to nil? If not I'm having a little trouble understanding when this `EOF`s, I would expect the call to `r.f` above to handle `EOF` so that `f` could signal when it has nothing left to return.

Reply:
`r.f` will be nil if the reader was created without a `ReaderFunc`. This pseudocode might help explain what this is doing:
```
If at the end of the data references in the current range:
  If there is a reader function, attempt to get another set of data references to read:
    If this succeeds: continue reading with the next set of data references.
    Else: return the error (can be EOF when done).
  Else:
    EOF
```
I think there may be some confusion about the role of the reader function.